File: src/orchestrator/intent_extractor.py
import os
import json
import logging
from typing import Any, Dict, List

import json5
import openai
from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type

from src.rag.pinecone_client import PineconeClient

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.INFO)

class IntentExtractionError(Exception):
    pass

class IntentExtractor:
    """
    IntentExtractor drives the LLM-based extraction of a canonical intent object
    from a raw requirement string, optionally enriching with RAG-based context.
    """
    def __init__(self,
                 pinecone_client: PineconeClient = None,
                 prompt_path: str = None):
        self.pinecone_client = pinecone_client or PineconeClient()
        self.prompt_path = prompt_path or os.getenv("INTENT_PROMPT_PATH", "config/prompts/intent.json5")
        self.prompt_config = self._load_prompt_template()
        openai.api_key = os.getenv("LLM_API_KEY")

    def _load_prompt_template(self) -> Dict[str, Any]:
        try:
            with open(self.prompt_path, "r") as f:
                return json5.load(f)
        except Exception as e:
            logger.error(f"Failed to load intent prompt template from {self.prompt_path}: {e}")
            raise IntentExtractionError(f"Prompt load failure: {e}")

    def _preprocess(self, text: str) -> str:
        # simple normalization: trim whitespace; extendable for markup stripping
        return text.strip()

    def _retrieve_context(self, request_id: str, top_k: int = 3) -> List[str]:
        try:
            return self.pinecone_client.query_context(request_id, top_k=top_k)
        except Exception as e:
            logger.warning(f"Context retrieval failed (continuing without RAG): {e}")
            return []

    @retry(retry=retry_if_exception_type(openai.error.RateLimitError),
           wait=wait_exponential(multiplier=1, min=1, max=10),
           stop=stop_after_attempt(3))
    def _call_llm(self, messages: List[Dict[str, str]], model: str, temperature: float, max_tokens: int) -> str:
        try:
            resp = openai.ChatCompletion.create(
                model=model,
                messages=messages,
                temperature=temperature,
                max_tokens=max_tokens
            )
            return resp.choices[0].message.content
        except Exception as e:
            logger.error(f"LLM call failed: {e}")
            raise

    def extract_intent(self, requirement: str, request_id: str) -> Dict[str, Any]:
        """
        Main entrypoint: returns a dict matching the canonical intent schema.
        """
        cleaned = self._preprocess(requirement)
        contexts = self._retrieve_context(request_id)
        system_msg = self.prompt_config.get("system_instructions", "You are a helpful assistant.")
        template = self.prompt_config.get("template", "")
        try:
            prompt_text = template.format(requirement=cleaned, context="\n".join(contexts))
        except Exception as e:
            logger.error(f"Prompt formatting failed: {e}")
            raise IntentExtractionError(f"Prompt formatting error: {e}")

        messages = [
            {"role": "system", "content": system_msg},
            {"role": "user", "content": prompt_text}
        ]

        model = self.prompt_config.get("model", "gpt-4")
        temperature = self.prompt_config.get("temperature", 0.0)
        max_tokens = self.prompt_config.get("max_tokens", 512)

        try:
            raw_output = self._call_llm(messages, model, temperature, max_tokens)
        except Exception as e:
            raise IntentExtractionError(f"LLM extraction failed: {e}")

        try:
            intent_obj = json.loads(raw_output)
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse JSON from LLM output: {e}\nOutput was: {raw_output}")
            raise IntentExtractionError(f"Invalid JSON output: {e}")

        # Optionally attach metadata
        metadata = {
            "request_id": request_id,
            "raw_response": raw_output
        }
        return {"intent": intent_obj, "metadata": metadata}

File: src/agent/steps/intentExtractor.ts
import { llmClient } from '../../services/llmClient'
import config from 'config'

export interface IntentExtractionResult {
  intent: string
  objectives: string[]
  constraints: string[]
}

/**
 * Extracts user intent, objectives, and constraints from a raw requirement string.
 * Sends a structured prompt to the LLM and parses the JSON response.
 * @param requirement - The raw requirement text to analyze.
 * @returns A promise resolving to the extracted intent object.
 * @throws Error if the LLM call fails or the response cannot be parsed/validated.
 */
export async function extractIntent(requirement: string): Promise<IntentExtractionResult> {
  const llmParams = config.get<Record<string, any>>('llm.intentExtractor')
  const systemPrompt = 'You are an AI assistant that analyzes user requirements.'
  const userPrompt = `
Extract the user's intent, objectives, and constraints from the following requirement.  
Respond ONLY with a JSON object with the keys "intent" (string), "objectives" (array of strings), and "constraints" (array of strings).

Requirement:
"""${requirement}"""
`

  try {
    const response = await llmClient.chatCompletion({
      model: llmParams.model,
      temperature: llmParams.temperature,
      max_tokens: llmParams.maxTokens,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userPrompt.trim() }
      ]
    })

    const message = response.choices?.[0]?.message?.content
    if (!message) {
      throw new Error('Empty response from LLM for intent extraction')
    }

    let parsed: any
    try {
      parsed = JSON.parse(message)
    } catch (err) {
      throw new Error(`Failed to parse JSON from LLM response: ${(err as Error).message}`)
    }

    if (
      typeof parsed.intent !== 'string' ||
      !Array.isArray(parsed.objectives) ||
      !Array.isArray(parsed.constraints) ||
      !parsed.objectives.every(o => typeof o === 'string') ||
      !parsed.constraints.every(c => typeof c === 'string')
    ) {
      throw new Error(`Invalid intent extraction schema: ${JSON.stringify(parsed)}`)
    }

    return {
      intent: parsed.intent,
      objectives: parsed.objectives,
      constraints: parsed.constraints
    }
  } catch (error) {
    const err = error as Error
    console.error('Error in extractIntent:', err.message)
    throw err
  }
}